### ISSUE TITLE
"Import" suggestions should be before "Change spelling" suggestions

### ISSUE THREAD

**Issue by lstkz** [microsoft/TypeScript#38743]
I face the following situation multiple times per day.
- Type some code.
- Variable is not defined/imported.
- Press ctrl+enter to trigger the quick fixes menu.
- "Change spelling" is always the first suggestion.
- Press down arrow and select the import suggestion.

![image](https://user-images.githubusercontent.com/4478684/82725032-8bc57a80-9cda-11ea-8f60-a845bd77a232.png)

**TypeScript Version:**  3.9.2, 4.0.0-dev

**Search Terms:**
Import, suggestion, spelling

I tried to google it, but I couldn't find any solutions.

**Code**

See the screenshot above.

**Expected behavior:**
The import suggestion should be the first entry in quick fixes.

**Actual behavior:**
The import suggestion is the last entry in quick fixes.

**Related Issues:** 
#23258 I think it broke it. Also please see the last comment, @krryan mentions the same problem.

**Comment by lstkz** [comment]
Another example. I must press unnecessary 'down arrow' multiple times.

![image](https://user-images.githubusercontent.com/4478684/82725608-57ec5400-9cde-11ea-8566-2a92bbeab9f9.png)

**Comment by DanielRosenwasser** [comment]
I guess if you've got an exact match, that is more desirable.

Side note @mjbvz, I wonder if there's a way to group some of these "fix all"s better so that the list doesn't get too long.

**Comment by DanielRosenwasser** [comment]
The fix should simply be a matter of reordering when each codefix gets registered by updating the `tsconfig.json` in `services/`

**Comment by mjbvz** [comment]
@DanielRosenwasser Opened https://github.com/microsoft/vscode/issues/98613 for 'add all missing imports' coming in the middle of the other import actions

**Comment by Gum-Joe** [comment]
@DanielRosenwasser if it's that simple to fix I'm happy to submit a PR here? Or is there something I am missing, given @mjbvz's comments?

**Comment by arunaaabh95** [comment]
Hi @Gum-Joe ,
wanted to give a head's up. I can see that in ```services/tsconfig.json``` we are already calling ```importFixes.ts``` before ```fixSpelling.ts``` 
I also changed the test case ```CodeFixSpellingVsImport.ts``` to check if **import fix** is being called before **change spelling** suggestion and I see it passing(it should not according to current behavior)

This test case is similar to what op has posted
```
/// <reference path='fourslash.ts' />

// Tests that the import fix is returned first.

// @Filename: /abc/a.ts
////export const ChallengeLoader = 0;

// @Filename: /b.ts
////const ChallengeHeader = 0;
////ChallengeLoader;

goTo.file("/b.ts");
verify.codeFixAvailable([
    { description: `Import 'ChallengeLoader' from module "./abc/a"` },
    { description: "Change spelling to 'ChallengeHeader'" },
]);
```
I might be missing something here. 

**TLDR:** We are missing something here, task needs more investigation

**Comment by Gum-Joe** [comment]
Thank for the head's up @arunaaabh95 
So if this is fixed, presumably it's in TS 3.9, which is bundled in VSCode.
Yet, having just checked this myself, the issue is still there.
My assumption at the moment is that there is some code in VSCode changing the order of fixes.
A few avenues to explore:

- Is there another IDE we can test with (besides VSCode Insiders, which might be worth testing)? Does Atom or IntelliJ use the same TS services?
- Track down how the suggestions are generated and walk through the code. Probably some good old fashioned `console.log()`ing?
   - We're probably looking at creating some custom VSCode release using a modified version of TS? This could take a while.

Also, where is `CodeFixSpellingVsImport.ts` BTW?

**Comment by arunaaabh95** [comment]
Update: After reading this similar issue [#27614](https://github.com/microsoft/TypeScript/issues/27614). These changes need to be handled on the editor side

**Comment by arunaaabh95** [comment]
@Gum-Joe 
you can find ```CodeFixSpellingVsImport.ts``` in ```test/cases/fourslahses``` directory

**Comment by Gum-Joe** [comment]
Thank you @arunaaabh95. I'll probably have a look tonight or at some point myself. It seems the fix for this is quite large and could be a major refactor? I agree with the above comment saying this is not a good first issue, since the fix looks quite big? (still going to work/help with it though)

**Comment by arunaaabh95** [comment]
@Gum-Joe I think @DanielRosenwasser and @mjbvz  are better judge of this, 
I started looking at this issue today itself and this is the first time I have investigated this code base but opening an issue in vscode seems to the right direction to me

**Comment by zachkirsch** [comment]
Glad to see there's already an issue files for this! Any update here?

**Comment by snigui** [comment]
changed the code error number in diagnosticsMessages.json just in case to make sure import comes before check spelling. Don't think this is an issue within TypeScript, but more so with VS Code's suggestions ordering?

**Comment by sandersn** [comment]
@mjbvz can you give us an idea of what might be re-ordering suggestions? Emacs has the correct order so I think that eliminates the language service as the cause of this.
